### ISSUE TITLE
Geometric Pooling layer between 2 graph convolution

### ISSUE THREAD

**Issue by yorard123** [rusty1s/pytorch_geometric#188]
Hi,

After all I would like to congrats you, I think you are doing an amazing job. This library is very complete :)

Maybe it is a silly question (I'm sorry if it is). I was wondering if it is possible to put a geometric pooling layer between two edge-conditioned convolution as it is done by Martin Simonovsky in 
Dynamic Edge-Conditioned Filters in Convolutional Neural Networks on Graphs. Suppose that you have a point cloud as input, you use your transformation to convert this point cloud to a graph and use the dynamic edge-conditioned convolution. After this operation you want to downsampling the output using a geometric max pooling using the Voxel Grid Pooling. Is it possible to do this? The question is because I was checking pytorch_cluster library and I did not see any backward method implemented.  Then far as I understond if I use this layer in the middle of a network the gradients will not propagate properly, right? Am i missing something? 

Thank you!

**Comment by rusty1s** [comment]
Hi, `torch_cluster` implements with some deterministic clustering algorithms for point clouds and graphs. Since graphs are discrete objects, a discrete clustering does not support differentiation and hence `torch_cluster` does not implement backward passes. However, we can still backpropagate gradients through `x` and `pos` based on the resulting `cluster` vector. This is done in `max_pool` and `max_pool_x`.

You can see this in action in the MNIST examples (graclus and voxel grid). You can simply replace the SplineConv with EdgeConv to implement your idea.

**Comment by yorard123** [comment]
Hi,
Understood the concept. I was checking the max_ppol and max_pool_x and there is not a explicit backpropagation, right? Is it using autograd func of pytorch?

Reviewing the EdgeConv I realize that is not the same operation that it is described in Dynamic Edge-Conditioned Filters. I started the implementation of this conv using the `MessagePassing class` But I am a little bit stuck understanding some of the concepts of the class. The formulation is that: (knn includes self loops)

![equation](http://latex.codecogs.com/gif.latex?X_i%5El%20%3D%20%5Cfrac%7B1%7D%7B%7CN%7C%7D%20%5Csum_%7Bj%5Cin%20N%28i%29%7D%20%5CTheta%5El_%7Bji%7D%20X_j%5E%7Bl-1%7D%20&plus;%20b%5El)

Where:
![equation2](http://latex.codecogs.com/gif.latex?%5CTheta)

Is the output of a mlp layer, that use the edge attribs to generate the weights used on the convolution.(each neighborhood has its own weights)

Then I am not sure how to fit this formula on the `MessagePassing class`. I suppose first thing that I need to do is calculate the weights(using the mlp) on the forward step. But then, how can I propagate this weights to the message method in order to do the multiplication inside the avg aggregation?
Maybe, I need to do all the calculations inside the forward. I suppose that it will have an impact on the performance, right?

Thanks!

**Comment by rusty1s** [comment]
Hi, yes max pool uses autograd capabilities of PyTorch (and `torch-scatter`).

Actually, our [`EdgeConv`](https://rusty1s.github.io/pytorch_geometric/build/html/modules/nn.html#torch_geometric.nn.conv.EdgeConv) refers to [this](https://arxiv.org/abs/1801.07829). The `EdgeConv` from the Dynamic Edge-Conditioned Filters is named [`NNConv`](https://rusty1s.github.io/pytorch_geometric/build/html/modules/nn.html#torch_geometric.nn.conv.NNConv), which is the same Gilmer et al. proposed in their Message Passing paper. You should check its source code.

**Comment by yorard123** [comment]
To be the same as  Dynamic Edge-Conditioned Filters,  the `NNConv` needs to put `root_weight` to `false` and add self lops, right? Moreover, I am supposing that the parameter `pseudo` that forward requires is the `edge_attrib`, right? 
Now I am just curious, in your implementation, you are not using the `MessagePassing class`. Why not? In which cases is most efficient your `MessagePassing class`?

**Comment by rusty1s** [comment]
Yes you are right. The `MessagePassing` base class is relatively new, and I haven't updated all operators to use it. This is on my todo list. Efficiency should be about the same.

**Comment by amosella** [comment]
I see, thank you so much! For my side we can close the issue.

**Comment by rusty1s** [comment]
No problem :)
